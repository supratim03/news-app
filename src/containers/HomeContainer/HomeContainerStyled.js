import styled from "styled-components"

export const Container = styled.div`
    margin-top: 12%;
    margin-left: 100px;
    margin-right: 100px;
    @media (max-width: 768px) {
        margin-top: 35%;
        margin-left: 10px;
        margin-right: 10px;
    }
`;

export const GridBlock = styled.div`
    display: grid;
    grid-template-columns: repeat(2,1fr);
    width: 100%;
`;

export const ContainerHeaderBlock = styled.span`
    font-weight: bold;
    font-size: 24px;
    color: #000000;
`;

export const SelectBlock = styled.div`
    width: 100%;
    float: right;
    text-align: right;
`;

export const GridContainer = styled.div`
    display: grid;
    grid-gap: 25px;
    /* background-color: #2196F3; */
    padding: 10px;
    width: 100%;
    margin-top: 30px;
    grid-template-columns: 1fr 1fr 1fr;
    grid-auto-rows: minmax(100px, auto);
    @media (max-width: 768px) {
        grid-template-columns: repeat(2, 1fr); 
        width: auto;
    }
`;

export const CardBlock = styled.div`
    padding-bottom: 200px;
    min-height: 100vh;
    margin-left: 100px;
    margin-right: 100px;
    @media (max-width: 768px) {
        margin-left: 10px;
        margin-right: 10px;
    }
`;

export const Card = styled.div`
    background-repeat: no-repeat !important;
    background-size: cover !important;
    position: relative;
    cursor: pointer;
`;