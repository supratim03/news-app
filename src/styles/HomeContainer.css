.item-card0 {
    grid-row: 1/4;
}

.item-card1 {
    grid-row: 1/3;
}

.item-card2 {
    grid-row: 1/3;
}

.item-card3, .item-card4 {
    background-color: #0d3779 !important;
}

.card-row-2 {
    height: 300px;
}

.blank-img {
    content: url('../../public/Logo-white.png');
}

.card-container {
    padding: 20px;
}

.card-desc {
    position: absolute;
    bottom: 0;
    background-color: #0d3779;
    min-width: 100%;
    min-height: 50px;
    opacity: 0.8;
}

.title {
    color: white;
    font-weight: bold;
    font-size: 16px;
    font-family: 'Times New Roman', Times, serif;
}

@media (max-width: 768px) {
    .item-card2 {
        grid-row: auto
    }
}

.bookmark-card {
    height: 300px;
    background-repeat: no-repeat;
    background-size: cover;
}