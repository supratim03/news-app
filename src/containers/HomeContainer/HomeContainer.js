import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import FooterComponent from "../../components/FooterComponent/FooterComponent";
import HeaderComponent from "../../components/HeaderComponent/HeaderComponent";
import SelectComponent from "../../components/SelectComponent/SelectComponent";
import { getDefaultNews } from "../../store/actions/actions";
import { Container, GridBlock, ContainerHeaderBlock, SelectBlock, CardBlock, GridContainer, Card } from './HomeContainerStyled';
import '../../styles/HomeContainer.css';
import { useNavigate } from "react-router-dom";
import LoaderComponent from "../../components/LoaderComponent/LoaderComponent";

const HomeContainer = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate()
    const isLoading = useSelector(state => state.news.isLoading);
    const newsList = useSelector(state => state.news.defaultNews);
    const [showSearch, setShowSearch] = React.useState(false)

    useEffect(() => {
        dispatch(getDefaultNews(10, 'newest'))
    }, [])
    
    const handleArticleById = (id, url) => {
        sessionStorage.setItem('apiUrl', url);
        let value = id.replaceAll('/', '-')
        navigate(`/article/${value}`)
    }

    const prepareNewsGrid = newsList.map((news, index) => {
        let classValue = null;
        if (index >= 5) {
            classValue = `item-card${index} card-row-2`
        } else {
            classValue = `item-card${index}`
        }

        return (
            <Card
                key={news.id}
                style={index === 3 || index === 4 || news.fields?.thumbnail == null ? { backgroundColor: '#2455a1' } : { backgroundImage: "url(" + news.fields?.thumbnail + ")" }}
                className={classValue}
                onClick={() => handleArticleById(news.id, news.apiUrl)}
            >
                {index === 3 || index === 4 ? '' : <div style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100%' }}><div>
                    <div className={news.fields?.thumbnail == null ? "blank-img" : ""}></div>
                </div></div>}
                <div className={index === 3 || index === 4 ? "card-container" : "card-desc"}>
                    <div style={{ padding: '10px' }}>
                        <span className="title">
                            {news.webTitle}
                        </span>
                    </div>

                </div>

            </Card>

        )
    })

    const detectBodyClick = () => {
        setShowSearch(false)
    }

    const handleOnFocus = (value) => {
        setShowSearch(value)
    }

    return (
        <>
            <HeaderComponent showSearch={showSearch} onFocus={handleOnFocus} />
            <Container onClick={detectBodyClick}>
                <GridBlock>
                    <ContainerHeaderBlock>
                        {'Top Stories'}
                    </ContainerHeaderBlock>
                    <SelectBlock>
                        <SelectComponent />
                    </SelectBlock>
                </GridBlock>
            </Container>
            <CardBlock>
                {!isLoading ? <GridContainer>
                    {prepareNewsGrid}
                </GridContainer> : <LoaderComponent />}
            </CardBlock>
            {!isLoading && <FooterComponent />}
        </>

    )
}

export default HomeContainer;