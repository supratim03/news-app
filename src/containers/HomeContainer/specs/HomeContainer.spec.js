import React from "react";
import { mount } from "enzyme/build";
import configureStore from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import HomeContainer from "../HomeContainer";
import { Card } from "../HomeContainerStyled";

// mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
   useNavigate: () => mockedUsedNavigate
}));

describe("Home container", () => {
    const sagaMiddleWare = createSagaMiddleware();
    const mockStore = configureStore([sagaMiddleWare]);

    const initialState = {
        "news": {
            "defaultNews": [],
            "isLoading": true
        }
    };
    const store = mockStore(initialState);
    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <HomeContainer />
                </MemoryRouter>
            </Provider>
        );
    });

    it("render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });
})

describe("Home container with laoder", () => {
    const sagaMiddleWare = createSagaMiddleware();
    const mockStore = configureStore([sagaMiddleWare]);

    const initialState = {
        "news": {
            "defaultNews": [
                {
                  "id": "world/live/2023/feb/12/turkey-syria-earthquake-latest-news-updates-death-toll",
                  "type": "liveblog",
                  "sectionId": "world",
                  "sectionName": "World news",
                  "webPublicationDate": "2023-02-12T11:10:40Z",
                  "webTitle": "Turkey-Syria earthquake: death toll passes 28,000 as UN warns it could ‘more than double’ – latest news",
                  "webUrl": "https://www.theguardian.com/world/live/2023/feb/12/turkey-syria-earthquake-latest-news-updates-death-toll",
                  "apiUrl": "https://content.guardianapis.com/world/live/2023/feb/12/turkey-syria-earthquake-latest-news-updates-death-toll",
                  "fields": {
                    "thumbnail": "https://media.guim.co.uk/b051c775f96ce352a176956872189ca4688abb8c/0_544_8256_4954/500.jpg"
                  },
                  "isHosted": false,
                  "pillarId": "pillar/news",
                  "pillarName": "News"
                },
                {
                  "id": "world/live/2023/feb/12/russia-ukraine-war-latest-updates-ukraine-forces-hold-frontline-in-donetsk-kyivs-top-military-commander-says",
                  "type": "liveblog",
                  "sectionId": "world",
                  "sectionName": "World news",
                  "webPublicationDate": "2023-02-12T11:04:33Z",
                  "webTitle": "Russia-Ukraine war live: highest Russian casualties since first week of invasion, UK says; Poland casts doubt over jets to Kyiv",
                  "webUrl": "https://www.theguardian.com/world/live/2023/feb/12/russia-ukraine-war-latest-updates-ukraine-forces-hold-frontline-in-donetsk-kyivs-top-military-commander-says",
                  "apiUrl": "https://content.guardianapis.com/world/live/2023/feb/12/russia-ukraine-war-latest-updates-ukraine-forces-hold-frontline-in-donetsk-kyivs-top-military-commander-says",
                  "fields": {
                    "thumbnail": "https://media.guim.co.uk/07a6fe944698a60be27b4fa4ff48444499b92570/304_527_5435_3261/500.jpg"
                  },
                  "isHosted": false,
                  "pillarId": "pillar/news",
                  "pillarName": "News"
                },
                {
                  "id": "world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision",
                  "type": "article",
                  "sectionId": "world",
                  "sectionName": "World news",
                  "webPublicationDate": "2023-02-12T09:30:20Z",
                  "webTitle": "The BBC’s Steve Rosenberg: ‘The increasing aggression in Russia worries me – it could get bumpy’",
                  "webUrl": "https://www.theguardian.com/world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision",
                  "apiUrl": "https://content.guardianapis.com/world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision",
                  "fields": {
                    "thumbnail": "https://media.guim.co.uk/22309f2ae5c62b09c9490c1bbbd216492f9a1107/262_0_7729_4640/500.jpg"
                  },
                  "isHosted": false,
                  "pillarId": "pillar/news",
                  "pillarName": "News"
                },
              ],
            "isLoading": false
        }
    };
    const store = mockStore(initialState);
    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <HomeContainer />
                </MemoryRouter>
            </Provider>
        );
    });

    it("render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should navigate on click of card', () => {
        const element = wrapper.find(Card).at(0);
        element.simulate('click', "world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision", "https://www.theguardian.com/world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision");
        setTimeout(() => {
            expect(mockedUsedNavigate).toHaveBeenCalled();
        }, 0)
    })
})