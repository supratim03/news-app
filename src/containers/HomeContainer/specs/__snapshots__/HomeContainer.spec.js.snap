// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home container render correctly 1`] = `ReactWrapper {}`;

exports[`Home container with laoder render correctly 1`] = `ReactWrapper {}`;
